program/src/state.rs

```diff
@@ -3,58 +3,546 @@
 use crate::util::serialize_b58;
 use crate::{
-    account_map::{AccountMap, AccountSet, EntryConstantSize, PubkeyAndEntry},
     MINIMUM_STAKE_ACCOUNT_BALANCE, MINT_AUTHORITY, RESERVE_ACCOUNT, STAKE_AUTHORITY,
+    VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT,
 };
-use crate::{VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT};
 
 pub const LIDO_VERSION: u8 = 1;
 
 /// Size of a serialized `Lido` struct excluding validators and maintainers.
 ///
 /// To update this, run the tests and replace the value here with the test output.
 pub const LIDO_CONSTANT_SIZE: usize = 353;
-pub const VALIDATOR_CONSTANT_SIZE: usize = 49;
 
-pub type Validators = AccountMap<Validator>;
+/// Types of list entries
+/// Uninitialized should always be a first enum field as it catches empty list data errors
+#[derive(Clone, Debug, PartialEq, Eq, BorshDeserialize, BorshSerialize, Serialize, BorshSchema)]
+pub enum ListEntryType {
+    /// If the account has not been initialized, the enum will be 0
+    Uninitialized,
+    Validator,
+    Maintainer,
+}
+
+impl Default for ListEntryType {
+    fn default() -> Self {
+        ListEntryType::Uninitialized
+    }
+}
+
+/// Storage list for accounts in the pool.
+/// It is used to serialize account list on stake pool initialization
+#[repr(C)]
+#[derive(
+    Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema, Serialize,
+)]
+pub struct AccountList<T> {
+    /// Data outside of the list, separated out for cheaper deserializations
+    #[serde(skip_serializing)]
+    pub header: ListHeader<T>,
+
+    /// List of accounts in the pool
+    pub entries: Vec<T>,
+}
+
+pub type ValidatorList = AccountList<Validator>;
+
+/// Maintainer list account
+///
+/// Maintainers are granted low security risk privileges. Maintainers are
+/// expected to run the maintenance daemon, that invokes the maintenance
+/// operations. These are gated on the signer being present in this set.
+/// In the future we plan to make maintenance operations callable by anybody.
+pub type MaintainerList = AccountList<Maintainer>;
+
+/// Helper type to deserialize just the start of AccountList
+#[repr(C)]
+#[derive(
+    Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema, Serialize,
+)]
+pub struct ListHeader<T> {
+    /// Maximum allowable number of elements
+    pub max_entries: u32,
+
+    pub entry_type: ListEntryType,
+
+    phantom: PhantomData<T>,
+}
+
+/// Generic element of a list
+pub trait ListEntry: Pack + Default + Clone + BorshSerialize + PartialEq + Debug {
+    const TYPE: ListEntryType;
+
+    fn new(pubkey: Pubkey) -> Self;
+    fn pubkey(&self) -> Pubkey;
+
+    /// Performs a very cheap comparison, for checking if entry
+    /// info matches the account address.
+    /// First PUBKEY_BYTES of a ListEntry data should be the account address
+    fn memcmp_pubkey(data: &[u8], pubkey: &[u8]) -> bool {
+        sol_memcmp(&data[..PUBKEY_BYTES], pubkey, PUBKEY_BYTES) == 0
+    }
```

Comment:
This looks like something that will be very error-prone to use. What if you pass in the wrong data or wrong pubkey array? I would on the existing `Eq` impl for `Pubkey`. There are no cases left where we need to do a linear scan and compare the pubkey of every entry, now that we provide indexes in the instructions, right? So performance of a single `Eq` of 32 bytes per instruction call, that should be negligible compared to the other authorization checks we already do.

Reply:
`BigVec` has `find` and `find_mut` that use `sol_memcmp` to efficiently search an element with given pubkey. It works much faster than using iterator with transmuting (already discussed it up the list). 
We can't do nothing if someone passes wrong data. Even Solana gives us access to raw data through `accounts.lido.data`. If someone misuses the data it's his fault. Actually `memcmp_pubkey` is used only by BigVec internally and is not is not a public function `fn memcmp_pubkey`. So it is OK.

Reply:
Aah, it’s for finding duplicates.

But still, do we need it? Do we run into the compute budget otherwise? This level of unsafety feels like a high price to pay for a sanity check like preventing duplicates. (I’m not saying that this particular implementation is unsafe, but there are only so many things you can do that need extremely careful review, until somebody misses something somewhere. The only way to confidently keep things secure, is to fight complexity, and try very very hard to keep things as simple as possible.)

Reply:
I agree, partially. The compute budget is reduced by half of it's current value.
Rust all over the std uses unsafe code, so not use rust? A little bit of tested unsafe code will not harm. By the way `fund` uses a single line of unsafe code at return `Some(unsafe { &*(current_slice.as_ptr() as *const T) })`. We know bpf is OK with unalignment reads. It's not a problem.
A bigger problem is to over complicate code.
Marinade [also](https://github.com/marinade-finance/liquid-staking-program/blob/a309057f1eb3413070846d34e8fd5d83e99dc1c6/programs/marinade-finance/src/state.rs#L83-L87) uses unsafe but not much. 
We have just 4 unsafe declarations which are all in a single file.
Even if we use `read_unaligned` it will still be unsafe by definition. By the way I used `read_unaligned` in `swap_remove` [here](https://github.com/lidofinance/solido/blob/387d93d99ed23c0f376444559f2ddead3c78ed7a/program/src/big_vec.rs#L78)